File: client/src/protocol/ratchet.ts
import { sha256 } from "@noble/hashes/sha2.js";
import { aeadDecrypt, aeadEncrypt, dh, freshNonce, freshX, kdf } from "./keys";
import { b64d, b64e, concat, fromUtf8, utf8 } from "./bytes";

const ZERO32 = new Uint8Array(32);

export type Header = { dh: string; n: number; pn: number };

export type RatchetState = {
  dhsSk: Uint8Array;
  dhsPk: Uint8Array;
  dhr: Uint8Array | null;
  rk: Uint8Array;
  cks: Uint8Array | null;
  ckr: Uint8Array | null;
  ns: number;
  nr: number;
  pn: number;
  skipped: Record<string, string>;
};

function kdfRk(rk: Uint8Array, dhOut: Uint8Array): { rk: Uint8Array; ck: Uint8Array } {
  const out = kdf(dhOut, rk, "bleep-rk-v1", 64);
  return { rk: out.slice(0, 32), ck: out.slice(32) };
}

function kdfCk(ck: Uint8Array): { ck: Uint8Array; mk: Uint8Array } {
  const out = kdf(ck, ZERO32, "bleep-ck-v1", 64);
  return { ck: out.slice(0, 32), mk: out.slice(32) };
}

function skipKey(dhPub: Uint8Array, n: number): string {
  return `${b64e(dhPub)}:${n}`;
}

export function initAlice(sk: Uint8Array, bobDhPk: Uint8Array): RatchetState {
  const dhs = freshX();
  const { rk, ck } = kdfRk(sk, dh(dhs.sk, bobDhPk));
  return {
    dhsSk: dhs.sk,
    dhsPk: dhs.pk,
    dhr: bobDhPk,
    rk,
    cks: ck,
    ckr: null,
    ns: 0,
    nr: 0,
    pn: 0,
    skipped: {},
  };
}

export function initBob(sk: Uint8Array, bobDhSk: Uint8Array, bobDhPk: Uint8Array): RatchetState {
  return {
    dhsSk: bobDhSk,
    dhsPk: bobDhPk,
    dhr: null,
    rk: sk,
    cks: null,
    ckr: null,
    ns: 0,
    nr: 0,
    pn: 0,
    skipped: {},
  };
}

function dhRatchet(state: RatchetState, theirPk: Uint8Array): void {
  state.pn = state.ns;
  state.ns = 0;
  state.nr = 0;
  state.dhr = theirPk;
  const recv = kdfRk(state.rk, dh(state.dhsSk, theirPk));
  state.rk = recv.rk;
  state.ckr = recv.ck;
  const dhs = freshX();
  state.dhsSk = dhs.sk;
  state.dhsPk = dhs.pk;
  const send = kdfRk(state.rk, dh(state.dhsSk, theirPk));
  state.rk = send.rk;
  state.cks = send.ck;
}

export function ratchetEncrypt(state: RatchetState, plaintext: Uint8Array): { header: Header; ciphertext: Uint8Array } {
  if (!state.cks) throw new Error("sending chain not ready");
  const next = kdfCk(state.cks);
  state.cks = next.ck;
  const header: Header = { dh: b64e(state.dhsPk), n: state.ns, pn: state.pn };
  state.ns += 1;
  const nonce = freshNonce();
  const ad = utf8(JSON.stringify(header));
  const ct = aeadEncrypt(next.mk, nonce, plaintext, ad);
  return { header, ciphertext: concat(nonce, ct) };
}

export function ratchetDecrypt(state: RatchetState, header: Header, ciphertext: Uint8Array): Uint8Array {
  const their = b64d(header.dh);
  const keyId = skipKey(their, header.n);
  if (state.skipped[keyId]) {
    const mk = b64d(state.skipped[keyId]);
    delete state.skipped[keyId];
    return open(mk, header, ciphertext);
  }
  if (!state.dhr || !eq(state.dhr, their)) {
    skipUntil(state, header.pn);
    dhRatchet(state, their);
  }
  skipUntil(state, header.n);
  if (!state.ckr) throw new Error("receiving chain not ready");
  const next = kdfCk(state.ckr);
  state.ckr = next.ck;
  state.nr += 1;
  return open(next.mk, header, ciphertext);
}

function skipUntil(state: RatchetState, until: number): void {
  if (!state.ckr || !state.dhr) return;
  if (until - state.nr > 200) throw new Error("too many skipped");
  while (state.nr < until) {
    const next = kdfCk(state.ckr);
    state.ckr = next.ck;
    state.skipped[skipKey(state.dhr, state.nr)] = b64e(next.mk);
    state.nr += 1;
  }
}

function open(mk: Uint8Array, header: Header, ciphertext: Uint8Array): Uint8Array {
  const nonce = ciphertext.slice(0, 12);
  const ct = ciphertext.slice(12);
  const ad = utf8(JSON.stringify(header));
  return aeadDecrypt(mk, nonce, ct, ad);
}

function eq(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let d = 0;
  for (let i = 0; i < a.length; i++) d |= a[i] ^ b[i];
  return d === 0;
}

export function x3dhAlice(
  aliceXSk: Uint8Array,
  ephSk: Uint8Array,
  bobIdX: Uint8Array,
  bobSpk: Uint8Array,
  bobOpk?: Uint8Array | null,
): Uint8Array {
  const dh1 = dh(aliceXSk, bobSpk);
  const dh2 = dh(ephSk, bobIdX);
  const dh3 = dh(ephSk, bobSpk);
  const parts = bobOpk ? concat(dh1, dh2, dh3, dh(ephSk, bobOpk)) : concat(dh1, dh2, dh3);
  return kdf(parts, ZERO32, "bleep-x3dh-v1", 32);
}

export function x3dhBob(
  bobXSk: Uint8Array,
  bobSpkSk: Uint8Array,
  aliceIdX: Uint8Array,
  aliceEph: Uint8Array,
  bobOpkSk?: Uint8Array | null,
): Uint8Array {
  const dh1 = dh(bobSpkSk, aliceIdX);
  const dh2 = dh(bobXSk, aliceEph);
  const dh3 = dh(bobSpkSk, aliceEph);
  const parts = bobOpkSk ? concat(dh1, dh2, dh3, dh(bobOpkSk, aliceEph)) : concat(dh1, dh2, dh3);
  return kdf(parts, ZERO32, "bleep-x3dh-v1", 32);
}

export function fingerprintSession(rk: Uint8Array): string {
  return b64e(sha256(rk)).slice(0, 12);
}

export function encodePt(obj: unknown): Uint8Array {
  return utf8(JSON.stringify(obj));
}

export function decodePt(buf: Uint8Array): unknown {
  return JSON.parse(fromUtf8(buf));
}

export type SerializedRatchet = {
  dhsSk: string;
  dhsPk: string;
  dhr: string | null;
  rk: string;
  cks: string | null;
  ckr: string | null;
  ns: number;
  nr: number;
  pn: number;
  skipped: Record<string, string>;
};

export function dumpRatchet(s: RatchetState): SerializedRatchet {
  return {
    dhsSk: b64e(s.dhsSk),
    dhsPk: b64e(s.dhsPk),
    dhr: s.dhr ? b64e(s.dhr) : null,
    rk: b64e(s.rk),
    cks: s.cks ? b64e(s.cks) : null,
    ckr: s.ckr ? b64e(s.ckr) : null,
    ns: s.ns,
    nr: s.nr,
    pn: s.pn,
    skipped: s.skipped,
  };
}

export function loadRatchet(s: SerializedRatchet): RatchetState {
  return {
    dhsSk: b64d(s.dhsSk),
    dhsPk: b64d(s.dhsPk),
    dhr: s.dhr ? b64d(s.dhr) : null,
    rk: b64d(s.rk),
    cks: s.cks ? b64d(s.cks) : null,
    ckr: s.ckr ? b64d(s.ckr) : null,
    ns: s.ns,
    nr: s.nr,
    pn: s.pn,
    skipped: s.skipped,
  };
}

File: android/app/src/main/java/za/bleep/personal/protocol/Ratchet.kt
package za.bleep.personal.protocol

import org.json.JSONObject
import org.json.JSONObject.NULL as JsonNull

data class Header(val dh: String, val n: Int, val pn: Int) {
    fun json(): JSONObject = JSONObject().put("dh", dh).put("n", n).put("pn", pn)
    fun aad(): String = """{"dh":"$dh","n":$n,"pn":$pn}"""
}

class RatchetState(
    var dhsSk: ByteArray,
    var dhsPk: ByteArray,
    var dhr: ByteArray?,
    var rk: ByteArray,
    var cks: ByteArray?,
    var ckr: ByteArray?,
    var ns: Int = 0,
    var nr: Int = 0,
    var pn: Int = 0,
    val skipped: MutableMap<String, String> = mutableMapOf(),
)

object Ratchet {
    private val ZERO = ByteArray(32)

    private fun kdfRk(rk: ByteArray, dhOut: ByteArray): Pair<ByteArray, ByteArray> {
        val out = Crypto.kdf(dhOut, rk, "bleep-rk-v1", 64)
        return out.copyOfRange(0, 32) to out.copyOfRange(32, 64)
    }

    private fun kdfCk(ck: ByteArray): Pair<ByteArray, ByteArray> {
        val out = Crypto.kdf(ck, ZERO, "bleep-ck-v1", 64)
        return out.copyOfRange(0, 32) to out.copyOfRange(32, 64)
    }

    fun initAlice(sk: ByteArray, bobDhPk: ByteArray): RatchetState {
        val dhs = Crypto.freshX()
        val (rk, ck) = kdfRk(sk, Crypto.dh(dhs.sk, bobDhPk))
        return RatchetState(dhs.sk, dhs.pk, bobDhPk, rk, ck, null)
    }

    fun initBob(sk: ByteArray, bobDhSk: ByteArray, bobDhPk: ByteArray): RatchetState =
        RatchetState(bobDhSk, bobDhPk, null, sk, null, null)

    fun x3dhAlice(aliceXSk: ByteArray, ephSk: ByteArray, bobIdX: ByteArray, bobSpk: ByteArray, bobOpk: ByteArray?): ByteArray {
        val dh1 = Crypto.dh(aliceXSk, bobSpk)
        val dh2 = Crypto.dh(ephSk, bobIdX)
        val dh3 = Crypto.dh(ephSk, bobSpk)
        val parts = if (bobOpk != null) Bytes.concat(dh1, dh2, dh3, Crypto.dh(ephSk, bobOpk)) else Bytes.concat(dh1, dh2, dh3)
        return Crypto.kdf(parts, ZERO, "bleep-x3dh-v1", 32)
    }

    fun x3dhBob(bobXSk: ByteArray, bobSpkSk: ByteArray, aliceIdX: ByteArray, aliceEph: ByteArray, bobOpkSk: ByteArray?): ByteArray {
        val dh1 = Crypto.dh(bobSpkSk, aliceIdX)
        val dh2 = Crypto.dh(bobXSk, aliceEph)
        val dh3 = Crypto.dh(bobSpkSk, aliceEph)
        val parts = if (bobOpkSk != null) Bytes.concat(dh1, dh2, dh3, Crypto.dh(bobOpkSk, aliceEph)) else Bytes.concat(dh1, dh2, dh3)
        return Crypto.kdf(parts, ZERO, "bleep-x3dh-v1", 32)
    }

    fun encrypt(state: RatchetState, plaintext: ByteArray): Pair<Header, ByteArray> {
        val cks = state.cks ?: error("sending chain not ready")
        val (nck, mk) = kdfCk(cks)
        state.cks = nck
        val header = Header(Bytes.b64e(state.dhsPk), state.ns, state.pn)
        state.ns += 1
        val nonce = Bytes.random(12)
        val ad = Bytes.utf8(header.aad())
        val ct = Crypto.aeadEncrypt(mk, nonce, plaintext, ad)
        return header to Bytes.concat(nonce, ct)
    }

    fun decrypt(state: RatchetState, header: Header, ciphertext: ByteArray): ByteArray {
        val their = Bytes.b64d(header.dh)
        val keyId = "${header.dh}:${header.n}"
        state.skipped.remove(keyId)?.let { mk ->
            return open(Bytes.b64d(mk), header, ciphertext)
        }
        if (state.dhr == null || !state.dhr.contentEquals(their)) {
            skipUntil(state, header.pn)
            dhRatchet(state, their)
        }
        skipUntil(state, header.n)
        val ckr = state.ckr ?: error("receiving chain not ready")
        val (nck, mk) = kdfCk(ckr)
        state.ckr = nck
        state.nr += 1
        return open(mk, header, ciphertext)
    }

    private fun dhRatchet(state: RatchetState, theirPk: ByteArray) {
        state.pn = state.ns
        state.ns = 0
        state.nr = 0
        state.dhr = theirPk
        val recv = kdfRk(state.rk, Crypto.dh(state.dhsSk, theirPk))
        state.rk = recv.first
        state.ckr = recv.second
        val dhs = Crypto.freshX()
        state.dhsSk = dhs.sk
        state.dhsPk = dhs.pk
        val send = kdfRk(state.rk, Crypto.dh(state.dhsSk, theirPk))
        state.rk = send.first
        state.cks = send.second
    }

    private fun skipUntil(state: RatchetState, until: Int) {
        val ckr = state.ckr ?: return
        val dhr = state.dhr ?: return
        require(until - state.nr <= 200) { "too many skipped" }
        var ck = ckr
        while (state.nr < until) {
            val (nck, mk) = kdfCk(ck)
            ck = nck
            state.skipped["${Bytes.b64e(dhr)}:${state.nr}"] = Bytes.b64e(mk)
            state.nr += 1
        }
        state.ckr = ck
    }

    fun dump(state: RatchetState): JSONObject = JSONObject()
        .put("dhsSk", Bytes.b64e(state.dhsSk))
        .put("dhsPk", Bytes.b64e(state.dhsPk))
        .put("dhr", state.dhr?.let { Bytes.b64e(it) } ?: JsonNull)
        .put("rk", Bytes.b64e(state.rk))
        .put("cks", state.cks?.let { Bytes.b64e(it) } ?: JsonNull)
        .put("ckr", state.ckr?.let { Bytes.b64e(it) } ?: JsonNull)
        .put("ns", state.ns)
        .put("nr", state.nr)
        .put("pn", state.pn)
        .put("skipped", JSONObject().also { o -> state.skipped.forEach { (k, v) -> o.put(k, v) } })

    fun load(o: JSONObject): RatchetState = RatchetState(
        dhsSk = Bytes.b64d(o.getString("dhsSk")),
        dhsPk = Bytes.b64d(o.getString("dhsPk")),
        dhr = if (o.isNull("dhr")) null else Bytes.b64d(o.getString("dhr")),
        rk = Bytes.b64d(o.getString("rk")),
        cks = if (o.isNull("cks")) null else Bytes.b64d(o.getString("cks")),
        ckr = if (o.isNull("ckr")) null else Bytes.b64d(o.getString("ckr")),
        ns = o.getInt("ns"),
        nr = o.getInt("nr"),
        pn = o.getInt("pn"),
        skipped = mutableMapOf<String, String>().also { m ->
            val sk = o.optJSONObject("skipped") ?: JSONObject()
            sk.keys().forEach { k -> m[k] = sk.getString(k) }
        },
    )

    private fun open(mk: ByteArray, header: Header, ciphertext: ByteArray): ByteArray {
        val nonce = ciphertext.copyOfRange(0, 12)
        val ct = ciphertext.copyOfRange(12, ciphertext.size)
        val ad = Bytes.utf8(header.aad())
        return Crypto.aeadDecrypt(mk, nonce, ct, ad)
    }
}

object Canonical {
    fun of(obj: JSONObject): String {
        val keys = obj.keys().asSequence().toMutableList().sorted()
        return keys.joinToString(",", "{", "}") { k ->
            "\"$k\":${value(obj.get(k))}"
        }
    }

    private fun value(v: Any): String = when (v) {
        is JSONObject -> of(v)
        is Number -> if (v.toDouble() == v.toLong().toDouble()) v.toLong().toString() else v.toString()
        is Boolean -> v.toString()
        JSONObject.NULL -> "null"
        else -> JSONObject.quote(v.toString())
    }
}
